### PR TITLE
treesitter: add runtime queries feature

### PR DESCRIPTION
**Pull request by vigoux**
Following https://github.com/neovim/neovim/issues/12619#issuecomment-656832733 I decided to upstream the `runtime query` feature from `nvim-treesitter`. This allows users to _source_ query files from a runtime path at `queries/{lang}`

I removed a bit of machinery from our implementation, but this allows two things :
 - User defined queries (located in the appropriate `.config` subdirs.
 - Base languages, that is, extending a given query using the queries of some base language (like `C++` using the `C` query).

This allows us to create runtime `.scm` files that are `sourced` by treesitter.

Also refactors a bit things to allow a bit more flexibility.

EDIT: Extra addition is out of the box treesitter highlighting of C (still opt in though), using a query defined in `nvim-treesitter`.

### CODE REVIEW

File: runtime/doc/lua.txt [hunk 1]
```diff
@@ -574,6 +574,15 @@ retained for the lifetime of a buffer but this is subject to change. A plugin
 should keep a reference to the parser object as long as it wants incremental
 updates.
 
+Required runtime files					*lua-treesitter-files*
```
Comment (tjdevries) [inline runtime/doc/lua.txt:577]: I worry about naming this section "lua-treesitter-files", since the files aren't just for lua.

I don't know that we always have to prefix everything with `lua` for tree-sitter stuff.

End users don't care what language we implemented it in.

File: runtime/doc/lua.txt [hunk 2]
```diff
@@ -574,6 +574,15 @@ retained for the lifetime of a buffer but this is subject to change. A plugin
...
+
+Treesitter requires some files to work properly :
+	- Parsers : this is the `.so` files created whan compiling parsers,
+	  they are searched under the `parsers` subdirectory.
+	- Queries : these are `.scm` files that are "sourced" when
+	  `vim.treesitter.get_query(lang, type)` is called. They are searched under the
+	  `queries/{lang}` subdirectory.
```
Comment (tjdevries) [inline runtime/doc/lua.txt:584]: This section should probably be a bit more formal.

Can you split it in into two different subsections: `tree-sitter-parsers` and `tree-sitter-queries` and give some info of both.

A simple example of a `query` would be good.

And also, do we ship with the `c` parser as a `parser.o` or something as well? We could give that as an example for the parser section.

File: runtime/doc/lua.txt [hunk 3]
```diff
@@ -678,10 +687,16 @@ supported match predicate is `eq?` (both comparing a capture against a string
 and two captures against each other).
 
 vim.treesitter.parse_query(lang, query)
-						*vim.treesitter.parse_query(()*
+						*vim.treesitter.parse_query()*
 	Parse the query as a string. (If the query is in a file, the caller
         should read the contents into a string before calling).
 
+vim.treesitter.get_query(lang, query_name)
```
Comment (tjdevries) [inline runtime/doc/lua.txt:694]: These usually go like `vim.treesitter.get_query({lang}, {query_name})`

And then below you can reference them the same way to make them look parameter style for the docs.

File: runtime/doc/lua.txt [hunk 4]
```diff
@@ -678,10 +687,16 @@ supported match predicate is `eq?` (both comparing a capture against a string
...
+						*vim.treesitter.get_query()*
+	Get the runtime query named `query_name`. This internally calls
+	`parse_query`. This will "source" all the runtime and user defined
```
Comment (tjdevries) [inline runtime/doc/lua.txt:697]: We should decide on a word for `"source"` and then use it without quotes throughout the document. It can be described in `tree-sitter-query` section

File: runtime/lua/vim/treesitter.lua [hunk 5]
```diff
@@ -44,50 +46,28 @@ function Parser:set_included_ranges(ranges)
   self.valid = false
 end
 
-local M = {
-  parse_query = vim._ts_parse_query,
-}
+-- TODO(vigoux): not that great way to do it, but that __index method bothers me...
+local M = vim.tbl_extend("error", query, lang)
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter.lua:50]: I'm not sure I understand the TODO / problem here.

File: runtime/lua/vim/treesitter.lua [hunk 6]
```diff
@@ -44,50 +46,28 @@ function Parser:set_included_ranges(ranges)
...
 
 setmetatable(M, {
   __index = function (t, k)
       if k == "TSHighlighter" then
-        t[k] = require'vim.tshighlighter'
+        t[k] = require'vim.treesitter.highlighter'
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter.lua:55]: I don't think `TSHighlighter` was around in 0.4.3, so we can probably remove this entirely and just return this for `highlighter` instead.

Alternatively, we can issue a "please use `treesitter.highlighter` when accessing highlights` message for awhile to let people switch and then change it afterwards.

File: runtime/lua/vim/treesitter/language.lua [hunk 7]
```diff
@@ -0,0 +1,26 @@
+local a = vim.api
+
+local M = {}
+
+function M.require_language(lang, path)
+  if vim._ts_has_language(lang) then
+    return true
+  end
+  if path == nil then
+    local fname = 'parser/' .. lang .. '.*'
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/language.lua:10]: I thought these had to have `.so` endings?

File: runtime/lua/vim/treesitter/language.lua [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    local paths = a.nvim_get_runtime_file(fname, false)
+    if #paths == 0 then
+      -- TODO(bfredl): help tag?
+      error("no parser for '"..lang.."' language")
+    end
+    path = paths[1]
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/language.lua:16]: This is probably fine, but we might want to at least let them know if there was more than one.

File: runtime/lua/vim/treesitter/query.lua [hunk 9]
```diff
@@ -0,0 +1,176 @@
+local a = vim.api
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:1]: Decent amount of spots in this function missing spaces between arguments to functions.

File: runtime/lua/vim/treesitter/query.lua [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+local function check_magic(str)
+  if string.len(str) < 2 or magic_prefixes[string.sub(str,1,2)] then
+    return str
+  end
+  return '\\v'..str
+end
+
+-- Some treesitter grammars extend others.
+-- We can use that to import the queries of the base language
+M.base_language_map = {
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:21]: Should have a function to add / modify these probably. Just in case we change the structure later.

Comment (tjdevries) [reply runtime/lua/vim/treesitter/query.lua:21]: And with that, then maybe hide this map as well (like, just make it a local variable and access it with some helper functions outside of the module).

File: runtime/lua/vim/treesitter/query.lua [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+  cpp = {'c'},
+  typescript = {'javascript'},
+  tsx = {'typescript', 'javascript'},
+}
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:25]: Additionally, this could probably use a test (you can resolve this if you added a test that uses this method somewhere below).

File: runtime/lua/vim/treesitter/query.lua [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+  lang.require_language(language)
+  local self = setmetatable({}, Query)
+  self.query = vim._ts_parse_query(language, vim.fn.escape(query,'\\'))
+  self.info = self.query:inspect()
+  self.captures = self.info.captures
+  self.regexes = {}
+  for id,preds in pairs(self.info.patterns) do
+    local regexes = {}
+    for i, pred in ipairs(preds) do
+      if (pred[1] == "match?" and type(pred[2]) == "number"
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:37]: Can you break this section into something like:

```lua
local predicate_handlers = {}
predicate_handlers["match?"] = function(predicate, other_required_params...)
  --   ... handling predicate stuff here
end
```
This way when we add more predicates it's obvious.

File: runtime/lua/vim/treesitter/query.lua [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+          and type(pred[3]) == "string") then
+        regexes[i] = vim.regex(check_magic(pred[3]))
+      end
+    end
+    if next(regexes) then
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:42]: This should probably be `if not vim.tbl_isempty` since it's more obvious.

File: runtime/lua/vim/treesitter/query.lua [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+  if not preds then
+    return true
+  end
+  local regexes = self.regexes[pattern]
+  for i, pred in pairs(preds) do
+    -- Here we only want to return if a predicate DOES NOT match, and
+    -- continue on the other case. This way unknown predicates will not be considered,
+    -- which allows some testing and easier user extensibility (#12173).
+    -- Also, tree-sitter strips the leading # from predicates for us.
+    if pred[1] == "eq?" then
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:70]: This should go along with the same idea that I had mentioned before with the predicate handlers. I don't want to have a huge if-else statement here. We can dispatch to the predicates if we have one using a table, and it will be much more obvious IMO what is happening.

Comment (tjdevries) [reply runtime/lua/vim/treesitter/query.lua:70]: Additionally, if we add a table method, we can allow users to add their own predicates if they want with `vim.treesitter.query.add_predicate(name, handler)` or whatever we decide, which is pretty cool. Does not have to be implemented immediately though.

File: runtime/lua/vim/treesitter/query.lua [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+      end
+      if not regexes[i]:match_line(bufnr, start_row, start_col, end_col) then
+        return false
+      end
+    end
+  end
+  return true
+end
+
+function Query:iter_captures(node, bufnr, start, stop)
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:103]: I know most of these weren't super documented before, but it'd be helpful to document what each of the public functions does at least.

File: runtime/doc/lua.txt (hunk 1 above)
Comment (vigoux) [reply runtime/doc/lua.txt:577]: Yeah you are right, I will move that to `treesitter-files`.

File: runtime/doc/lua.txt (hunk 2 above)
Comment (vigoux) [reply runtime/doc/lua.txt:584]: > And also, do we ship with the `c` parser as a `parser.o` or something as well? We could give that as an example for the parser section.

Hmm that's a good question, I think yes, but I will have to look at the runtime files.

File: runtime/lua/vim/treesitter.lua (hunk 5 above)
Comment (vigoux) [reply runtime/lua/vim/treesitter.lua:50]: In the end I fixed it, will remove it

File: runtime/lua/vim/treesitter/language.lua (hunk 7 above)
Comment (vigoux) [reply runtime/lua/vim/treesitter/language.lua:10]: Not sure about that part... I think both `.o` and `.so` work.

File: runtime/lua/vim/treesitter/query.lua (hunk 13 above)
Comment (vigoux) [reply runtime/lua/vim/treesitter/query.lua:42]: Yep, that is just some old code I need to tidy

File: runtime/lua/vim/treesitter/language.lua (hunk 7 above)
Comment (jamessan) [reply runtime/lua/vim/treesitter/language.lua:10]: Doesn't it depend on the platform? Unix-like systems will probably be one of those, but Windows and maybe Mac will be different.

**Comment by vigoux** [comment]
Okay, it looks like we have a fully functionnal `vim.treesitter.highlighter.new({bufnr}, {lang})` !
Still need to wrtie some doc and some tests to be sure everything is okay, but otherwise, I think we're good !

File: runtime/lua/vim/treesitter/query.lua (hunk 12 above)
Comment (vigoux) [reply runtime/lua/vim/treesitter/query.lua:50]: Here is a thought that I had for a while : regexes are a pain in current implementation, they leverage vim builtin regexes while the whole treesitter runtime is in lua. I really think that using lua regexes is better, and can even provide better performances (here again, just a gut feeling, but the JIT can help here). Plus removing this and using lua regexes will simply remove this part of the code, and simplify the creation of queries (to go full lua), and have more flexibility.

File: runtime/lua/vim/treesitter/language.lua (hunk 7 above)
Comment (tjdevries) [reply runtime/lua/vim/treesitter/language.lua:16]: OHh.... will this make `c` match `cpp`?

Comment (vigoux) [reply runtime/lua/vim/treesitter/language.lua:16]: Oh that's a good point.

Comment (tjdevries) [reply runtime/lua/vim/treesitter/language.lua:16]: It should probably be like `string.format("parser/%s\\..*", lang)`  or similar. It should be <language>.<extension>

Comment (vigoux) [reply runtime/lua/vim/treesitter/language.lua:16]: Confirmed to be in its correct form, this wil be expanded to 'parser/{mylang}.*` and at some point will be passed to `gen_expand_wildcards()` which only consider `*` as wildcards.

File: runtime/doc/lua.txt [hunk 16]
```diff
@@ -574,6 +574,26 @@ retained for the lifetime of a buffer but this is subject to change. A plugin
 should keep a reference to the parser object as long as it wants incremental
 updates.
 
+Parser files						*treesitter-parsers*
+
+Parsers are the heart of tree-sitter. They are shared objects (`.so`) that
+tree-sitter will search for in the `parsers` runtime directory.
+
+For a parser to be available for a given language, there must be a file named
+`{lang}.so` within the parser directory.
```
Comment (tjdevries) [inline runtime/doc/lua.txt:583]: Is what jamessan was saying applicable here? Can it be something besides `.so`? Don't know if it even matters.

File: runtime/doc/lua.txt [hunk 17]
```diff
@@ -574,6 +574,26 @@ retained for the lifetime of a buffer but this is subject to change. A plugin
...
+
+Query files						*treesitter-query*
+
+Queries are objects that tree-sitter uses to extract informations. They are
+built from a `scheme`-like language, and treesitter parses them to build the
+object.
+
+Treesitter also searches for query files within the `queries` directory. For a
+given language, treesitter will search the queries in the `queries/{lang}`. A
+query file is a `scheme` file (ending with `.scm`).
```
Comment (tjdevries) [inline runtime/doc/lua.txt:593]: so the query file must have `.scm` extension?

File: runtime/doc/lua.txt [hunk 18]
```diff
@@ -677,17 +697,23 @@ Tree-sitter queries are supported, with some limitations. Currently, the only
...
+						*vim.treesitter.parse_query()*
+	Parse {query} as a string. (If the query is in a file, the caller
         should read the contents into a string before calling).
 
-query:iter_captures(node, bufnr, start_row, end_row)
+vim.treesitter.get_query({lang}, {query_name})
+						*vim.treesitter.get_query()*
+	Get the runtime query named {query_name}. This internally calls
+	|vim.treesitter.parse_query()|. This will "source" all the runtime and
+	user defined queries (see |lua-treesitter-files|).
```
Comment (tjdevries) [inline runtime/doc/lua.txt:709]: Not sure you need to explain what it does internally. I also don't know what you mean by "all" the runtime and user defined queries. It doesn't just source the one that I have with `{query_name}`?

File: runtime/lua/vim/treesitter.lua [hunk 19]
```diff
@@ -38,61 +46,53 @@ function Parser:_on_lines(bufnr, changed_tick, start_row, old_stop_row, stop_row
...
-function M.create_parser(bufnr, lang, id)
-  M.require_language(lang)
+--- Creates a new parser.
+--
+-- It is not recommended to use this, use vim.treesitter.get_parser() instead.
+--
+-- @param bufnr the buffer the parser will be tied to
+-- @param language the language of the parser.
+-- @param id the id the parser will have
+function M.create_parser(bufnr, language, id)
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter.lua:81]: If we don't want people to use this, can we just prefix it with an underscore then?

File: runtime/lua/vim/treesitter.lua [hunk 20]
```diff
@@ -108,6 +108,20 @@ function M.create_parser(bufnr, lang, id)
   return self
 end
 
+--- Gets the parser for this bufnr / ft combination.
+--
+-- If needed this will create the parser.
+-- In any case, the provided callbacks will be attached to the buffer.
+--
+-- @param bufnr hte buffer the parser should be tied to
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter.lua:116]: typo `hte`

File: runtime/lua/vim/treesitter.lua [hunk 21]
```diff
@@ -108,6 +108,20 @@ function M.create_parser(bufnr, lang, id)
...
+-- @param ft the filetype of this parser
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter.lua:117]: This will be annoying, but can you start the sentence after the variable name with an upper-case letter.

`@param ft The filetype...`

This way it is easier to read, IMO

File: runtime/lua/vim/treesitter/highlighter.lua [hunk 22]
```diff
@@ -75,7 +109,18 @@ end
 function TSHighlighter:set_query(query)
   if type(query) == "string" then
     query = vim.treesitter.parse_query(self.parser.lang, query)
+  elseif query == nil then
+    query = vim.treesitter.get_query(self.parser.lang, 'highlights')
+
+    if query == nil then
+      a.err_writeln("No highlights.scm query found for " .. self.parser.lang)
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/highlighter.lua:116]: isn't this supposed to be `nvim_err_writeln`?

File: runtime/lua/vim/treesitter/highlighter.lua [hunk 23]
```diff
@@ -75,7 +109,18 @@ end
...
+
+      if query == nil then
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/highlighter.lua:118]: you check for nil twice here it looks like?

File: runtime/lua/vim/treesitter/language.lua [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+-- @param path optionnal path the parser is located at
+function M.require_language(lang, path)
+  if vim._ts_has_language(lang) then
+    return true
+  end
+  if path == nil then
+    local fname = 'parser/' .. lang .. '.*'
+    local paths = a.nvim_get_runtime_file(fname, false)
+    if #paths == 0 then
+      -- TODO(bfredl): help tag?
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/language.lua:19]: You can probably fix this todo by linking to the help section

File: runtime/doc/lua.txt (hunk 16 above)
Comment (vigoux) [reply runtime/doc/lua.txt:583]: I don't think it matters, but we are sure that it works this way and that's the way @bfredl tested it.

File: runtime/doc/lua.txt (hunk 17 above)
Comment (vigoux) [reply runtime/doc/lua.txt:593]: Yep

File: runtime/doc/lua.txt (hunk 18 above)
Comment (vigoux) [reply runtime/doc/lua.txt:709]: It actually will concatenate every query named `{query_name}` and parse that as the runtime query.

File: runtime/lua/vim/treesitter.lua [hunk 25]
```diff
@@ -108,6 +108,20 @@ function M.create_parser(bufnr, lang, id)
   return self
 end
 
+--- Gets the parser for this bufnr / ft combination.
+--
+-- If needed this will create the parser.
+-- In any case, the provided callbacks will be attached to the buffer.
```
Comment (teto) [inline runtime/lua/vim/treesitter.lua:114]: what if there are already callbacks ? `unconditionnally attach the provided callback`

File: runtime/lua/vim/treesitter/highlighter.lua [hunk 26]
```diff
@@ -38,6 +68,7 @@ function TSHighlighter.new(query, bufnr, ft)
 
   local tree = self.parser:parse()
   self.root = tree:root()
+
```
Comment (teto) [inline runtime/lua/vim/treesitter/highlighter.lua:71]: newline

**Review by teto — COMMENTED** [review]
thanks for your commitment to improve treesitter.
- So for now we have `parser` and `query` subfolders. Can you foresee additional folders ? 
- can a user override the default C query by putting one in his own query/c.scm ?
- what's the status of the PR ? ready to merge ?

**Comment by vigoux** [comment]
>     * So for now we have `parser` and `query` subfolders. Can you foresee additional folders ?

I don't think we need more than these, with the things we have done within `nvim-treesitter` only those two things were useful, as the content of the `query` subfolder can hold queries with different _meaning_ (`highlights`, `definition`, `fold` or anything).

>     * can a user override the default C query by putting one in his own query/c.scm ?

Yes, this is possible as of today, and this for any `query_name`, just by putting something in `queries/c/{query_name}.scm`. One thing to note is that this is not overwriting things, but treesitter will search for each runtime file name `queries/c/{query_name}.scm` and concatenate them.

>     * what's the status of the PR ? ready to merge ?

IMO yes, but the query overwriting problem might be blocking.

File: runtime/lua/vim/treesitter.lua (hunk 25 above)
Comment (vigoux) [reply runtime/lua/vim/treesitter.lua:114]: Then the provided callbakc will be called _after_ the already attached callback.

File: runtime/doc/lua.txt (hunk 16 above)
Comment (teto) [reply runtime/doc/lua.txt:583]: `.so` is for linux, dylib for mac and dll for windows. does the code search for the correct extension ? 
maybe rephrase to `{lang}.{ext}  where {ext} is "so" on linux, dylib on mac, `

File: runtime/doc/lua.txt (hunk 17 above)
Comment (teto) [reply runtime/doc/lua.txt:593]: `queries/{lang}.scm` would be more consistent with the rest of the doc.

File: runtime/doc/lua.txt (hunk 18 above)
Comment (teto) [reply runtime/doc/lua.txt:709]: sorry I asked the same question at the bottom :$ I think it's important to describe since it is not how nvim operates in general.

File: runtime/doc/lua.txt (hunk 17 above)
Comment (vigoux) [reply runtime/doc/lua.txt:593]: That is not how it works, it is `queires/{Lang}/{query_name}.scm`.
I will document that better.

**Comment by tjdevries** [comment]
I will review some more today.

File: runtime/doc/lua.txt (hunk 18 above)
Comment (tjdevries) [reply runtime/doc/lua.txt:709]: Hmmm... yeah I'm not sure I like that all that much. Why do we smash them all together?

File: runtime/lua/vim/treesitter/query.lua [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+  return table.concat(contents, '\n')
+end
+
+--- Gets a fully fledged query from runtime files
+--
+-- @param lang The source language
+-- @param query_name The name of the query (without `.scm`)
+--
+-- @returns The query
+function M.get_query(language, query_name)
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter/query.lua:92]: perhaps query name should allow for a complete filename to be passed to it? So you can specify some random file as well? 

You can check if that file exists and then read from that one if it does.

This might solve @teto 's concern.

File: test/functional/lua/treesitter_spec.lua [hunk 28]
```diff
@@ -292,9 +309,9 @@ static int nlua_schedule(lua_State *const lstate)
     ]]}
 
     exec_lua([[
-      local TSHighlighter = vim.treesitter.TSHighlighter
+      local TSHighlighter = vim.treesitter.highlighter
```
Comment (tjdevries) [inline test/functional/lua/treesitter_spec.lua:312]: weird to still call this TSHighlighter? I'm not sure.

File: runtime/doc/lua.txt (hunk 18 above)
Comment (vigoux) [reply runtime/doc/lua.txt:709]: That's the way we did it in `nvim-treesitter` and it works well for highlighting. My worry though is for other kind of queries (i.e. locals or textobjects), because you can't overwrite part of the queries, but just add some other queries.

After some thinking, I think that this is a blocker for this PR, and we might want to find a way for users to both override and add new items to the queries. This echoes the following in the traking issue `consider a standard macro/template system for developing highlight queries (could involve fennel :)`. I actually agree that fennel could be a good idea, as it will let us write something like this (I don't know if that's valid fennel)
```fennel
(let [query {}] 
  (set query.function "((function) @test)" ;; Maybe a macro can be used here to avoid the quotes
)
```

Does this suit you ?

File: runtime/lua/vim/treesitter/query.lua (hunk 27 above)
Comment (vigoux) [reply runtime/lua/vim/treesitter/query.lua:92]: Should be good now, the `query_name` will be used as is if it is a readable file.

File: test/functional/lua/treesitter_spec.lua (hunk 28 above)
Comment (vigoux) [reply test/functional/lua/treesitter_spec.lua]: Renamed

File: runtime/lua/vim/treesitter.lua [hunk 29]
```diff
@@ -1,4 +1,6 @@
 local a = vim.api
+local query = require'vim.treesitter.query'
+local lang = require'vim.treesitter.language'
```
Comment (tjdevries) [inline runtime/lua/vim/treesitter.lua:3]: I don't really like having `lang` be the import from `language`.

1. it's confusing to shorten the module
2. it seems too easy to overwrite. Perhaps some more unique name would be better.

File: runtime/lua/vim/treesitter/query.lua (hunk 12 above)
Comment (theHamsta) [reply runtime/lua/vim/treesitter/query.lua:64]: Would be nice if the predicate handler could automatically define the negation of the predicate. Like `match?` could automatically define `not-match?`

Comment (theHamsta) [reply runtime/lua/vim/treesitter/query.lua:64]: Would be nice if the predicate handler could automatically define the negation of the predicate. Like `match?` could automatically define `not-match?`

Comment (theHamsta) [reply runtime/lua/vim/treesitter/query.lua:64]: Plus Lua regexes can be multi-line

Comment (vigoux) [reply runtime/lua/vim/treesitter/query.lua:64]: Agree on the negation part for predicates. Regexes will be tackled later.

File: runtime/lua/vim/treesitter.lua [hunk 30]
```diff
@@ -108,6 +108,20 @@ function M.create_parser(bufnr, lang, id)
   return self
 end
 
+--- Gets the parser for this bufnr / ft combination.
+--
+-- If needed this will create the parser.
+-- Unconditionnally attach the provided callback
```
Comment (brcolow) [inline runtime/lua/vim/treesitter.lua:114]: Typo:

Unconditionnally => Unconditionally

**Comment by vigoux** [comment]
As this PR also includes other features than just the runtime queries, I will separate this in two, so that the runtime queries is not blocking anymore for the other improvements made here.

**Comment by teto** [comment]
is it ok to merge if you rebase ?

**Comment by vigoux** [comment]
Not sure not sure, we made some improvements in `nvim-treesitter` leveraging the `after` directory.
This makes it more intuivite by saying that queries in the `after` directory will be used to extend "normal" (as in not in `after`).
I personally find it easier as you can both overwrite and extend by juste managing files, no weird comment magic involved or anything.

**Comment by vigoux** [comment]
Closed in favor of  #13059.
